R/labels.r

```diff
@@ -97,3 +101,111 @@ ylab <- function(label) {
 ggtitle <- function(label, subtitle = waiver()) {
   labs(title = label, subtitle = subtitle)
 }
+
+#' Extract an alt text from a plot
+#'
+#' This function returns a text that can be used as alt-text in webpages etc.
+#' It will either uses the `alt` label, added with `+ labs(alt = <...>)`, or
+#' synthesize one from the information in the plot itself. You can add a
+#' conclusion to the synthesized text using `+ labs(alt_insight = <...>)`.
+#'
+#' There is no way an automatically generated description can compete with one
+#' written by a human with knowledge of what the plot shows and in which
+#' context. We urge users to write their own alt text if at all possible.
+#' Guidance to how an effective alt-text is written can be found in
+#' [Writing Alt Text for Data Visualization](https://medium.com/nightingale/writing-alt-text-for-data-visualization-2a218ef43f81)
+#' and [Effective Practices for Description of Science Content within Digital Talking Books](https://www.wgbh.org/foundation/ncam/guidelines/effective-practices-for-description-of-science-content-within-digital-talking-books)
+#'
+#' @param p a ggplot object
+#' @param ... Currently ignored
+#'
+#' @return A text string
+#'
+#' @export
+#' @aliases alt_text
+#'
+#' @examples
+#' p <- ggplot(mpg, aes(displ, hwy)) +
+#'   geom_point()
+#'
+#' get_alt_text(p)
+#'
+#' p <- p + ggtitle("The relationship between displacement and yield in cars")
+#' get_alt_text(p)
+#'
+#' # It will use scale information if available
+#' p <- p + scale_x_continuous("highway miles per gallon")
+#' get_alt_text(p)
+#'
+#' # Add a short description of the main conclusion of the plot
+#' p <- p + labs(alt_insight = "The higher the yield, the lower the displacement")
+#' get_alt_text(p)
+#'
+#' # A user provided alt text takes precedence
+#' p <- p + labs(
+#'   alt = paste("A scatterplot showing the negative correlation between engine",
+#'               "displacement as a function of highway miles per gallon")
+#' )
+#'
+#' get_alt_text(p)
+#'
+get_alt_text <- function(p, ...) {
+  UseMethod("get_alt_text")
+}
+#' @export
+get_alt_text.ggplot <- function(p, ...) {
```

Comment:
Could we also get a method for the `ggplot_build` class? This would make the implementation for `shiny::renderPlot()` slightly easier, and might be useful in other situations

Reply:
Better yet, would it be sensible for `ggplot_build()` to call this function and tack it onto the result?